File: Kattis/BinaryExponentiation.hs
import Control.Arrow ((>>>))
import Data.Bits (shiftR)
import Data.Functor ((<&>))

main :: IO ()
main = do
    [a,e,m] <- getContents <&> (words >>> map read)
    print $ pow a e m

pow :: Integer -> Integer -> Integer -> Integer
pow _ 0 _ = 1
pow x n m | odd n     = (u'*x) `mod` m
          | otherwise = u'
    where 
        u = pow x (n `shiftR` 1) m
        u' = (u*u) `mod` m

File: Kattis/binaryexponentiation.py
def pow(x, n, m):
   if (n == 0): return 1
   u = pow(x,n//2,m)
   u = (u*u)%m
   if (n & 1): u = (u*x)%m
   return u

[a,e,m] = (int(i) for i in input().split())

print(pow(a,e,m))
